garden-backend-service/src/api/routes/modal/modal_apps.py

```diff
@@ -82,14 +68,86 @@ async def add_modal_app(
     model_dict["user_id"] = user.id
     model_dict["app_name"] = prefixed_app_name
     for modal_fn in model_dict["modal_functions"]:
-        modal_fn["hardware_spec"] = metadata["functions"][modal_fn["function_name"]]
+        name = modal_fn["function_name"]
+        modal_fn["hardware_spec"] = hardware_specs[name]
+
     modal_app_db_model = ModalApp.from_dict(model_dict)
 
     db.add(modal_app_db_model)
     await db.commit()
     return modal_app_db_model
 
 
+def _validate_modal_app_metadata(app_metadata: ModalAppCreateRequest):
+    """Validate that the metadata in the request matches what we parse from the modal file.
+
+    Args:
+        app_metadata: The app metadata from the request body
+
+    Raises:
+        ModalException: If there are inconsistencies between the metadata and parsed file
+    """
+    parse_results = parse_modal_file(app_metadata.file_contents)
+
+    if len(parse_results.apps) != 1:
+        raise ModalException(
+            detail=f"Invalid modal file: Found {len(parse_results.apps)} App objects. Expected exactly one.",
+            suggested_fix="Submit a file with a single modal App object.",
+        )
+
+    app_info = parse_results.apps[0]
+
+    # Validate app name matches
+    if app_info.app_name != app_metadata.app_name:
+        raise ModalException(
+            detail=f"App name mismatch: Got '{app_info.app_name}' from file but '{app_metadata.app_name}' in request.",
+            suggested_fix="Ensure the App name in your modal file matches the metadata.",
+        )
+
+    # Validate base image name
+    if app_info.image.base_image != app_metadata.base_image_name:
+        raise ModalException(
+            detail=f"Base image mismatch: Got '{app_info.image.base_image}' from file but '{app_metadata.base_image_name}' in request.",
+            suggested_fix="Ensure the base image in your modal file matches the metadata.",
+        )
+
+    # Check requirements match
+    request_pip_requirements = app_metadata.requirements
+    file_pip_requirements = app_info.image.pip_requirements
+    if set(request_pip_requirements) != set(file_pip_requirements):
+        logger.warning(
```

Comment:
tbh just wan't thinking too hard and demoted them from an exception to a logger warning. imo not worth passing on to the user, since they would have needed to actively change those from the pre-filled form values